controllers/topology/cluster_controller_test.go

```diff
@@ -0,0 +1,361 @@
+		WithInfrastructureClusterTemplate(infrastructureCluster).
+		WithControlPlaneTemplate(controlPlane).
+		WithControlPlaneInfrastructureMachineTemplate(infrastructureMachineTemplate).
+		WithWorkerMachineDeploymentClasses([]clusterv1.MachineDeploymentClass{*machineDeploymentClass}).
+		Build()
+	cluster := builder.Cluster(ns.Name, clusterName).
+		WithTopology(
+			builder.ClusterTopology(clusterClass.Name).
+				WithWorkerTopology(machineDeploymentTopology).
+				WithWorkerTopology(machineDeploymentTopology2).
+				WithVersion("1.22.2").
+				WithControlPlaneReplicas(3).
+				Build()).
+		Build()
+
+	// Create a set of initObjects from the objects created above to add to the API server when the test environment starts.
+	initObjs := []client.Object{
+		clusterClass,
+		cluster,
+		infrastructureCluster,
+		infrastructureMachineTemplate,
+		bootstrapTemplate,
+		controlPlane,
+	}
+
+	testName := "Cluster creation from barebones cluster object and clusterClass."
+	t.Run(testName, func(t *testing.T) {
+		defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterTopology, true)()
+
+		for _, obj := range initObjs {
+			err := env.Create(ctx, obj)
+			g.Expect(err).ToNot(HaveOccurred())
+		}
+		actualCluster := &clusterv1.Cluster{}
+		// Get Cluster object and check that relevant fields are set as expected.
+		g.Eventually(func() bool {
+			return assertClusterReconcile(ns.Name, clusterName, actualCluster)
+		}, timeout).Should(BeTrue())
+
+		// Check if InfrastructureCluster has been created and has the correct labels and annotations.
+		g.Eventually(func() bool {
+			return assertInfrastructureClusterReconcile(actualCluster)
+		}, timeout).Should(BeTrue())
+
+		// Check if ControlPlane has been created and has the correct version, replicas, labels and annotations.
+		g.Eventually(func() bool {
+			return assertControlPlaneReconcile(actualCluster, clusterClass)
+		}, timeout).Should(BeTrue())
+		g.Eventually(func() bool {
+			return assertMachineDeploymentsReconcile(actualCluster)
+		}, timeout).Should(BeTrue())
+	})
+}
+
+// assertClusterReconcile checks if the Cluster object:
+// 1) Has its InfrastructureReference and ControlPlane reference added correctly.
+// 2) InfrastructureReference and ControlPlaneRef have the expected Group, Version and Kind.
+func assertClusterReconcile(namespace, clusterName string, cluster *clusterv1.Cluster) bool {
+	key := client.ObjectKey{Name: clusterName, Namespace: namespace}
+	if err := env.Get(ctx, key, cluster); err != nil {
+		return false
+	}
+	// Check if relevant managed topology labels are present.
+	if ok := ensureClusterTopologyOwnedLabel(cluster); !ok {
+		return false
+	}
+	// Check if InfrastructureRef exists and is of the expected Kind and APIVersion.
+	if ok := referenceExistsWithCorrectKindAndAPIVersion(cluster.Spec.InfrastructureRef,
+		builder.GenericInfrastructureClusterKind,
+		builder.InfrastructureGroupVersion); !ok {
+		log.Infof("We failed on infraref %v", cluster)
```

Comment:
logs need to be updated across the tests here to use the topology logger properly - will do in a new commit.

Reply:
Can we use t.Log?

Reply:
It means we'd have to pass t along with the assert and other functions which would make it a little less readable. I can definitely do it, but I prefer a global magic logger for a test like this.

Reply:
But does it work the same way?
```bash
// Log formats its arguments using default formatting, analogous to Println,
// and records the text in the error log. For tests, the text will be printed only if
// the test fails or the -test.v flag is set. For benchmarks, the text is always
// printed to avoid having performance depend on the value of the -test.v flag.
```

Usually t.Log is intentionally used in unit tests to get that behaviour

Reply:
I think this is irrelevant now as this is no longer using the logger. I think the behaviour was equivalent, but I was only viewing the logs from inside Goland's Run funciton so I'm not certain it would be the same on CLI or other tools.

Reply:
Agree it's not relevant anymore.

Afaik the difference is that log.Info always logs and t.Log only if it fails. I'm also not sure if log.Info also includes it correctly in the junit report which is generated by ginkgo.